elastic/logstash-filter-elastic_integration
# Implemented version check with remote cluster

## Release notes

Add a version preflight check so that the plugin can connect only to Elasticsearch of the same  or previous version of the Ingest Node code bundled within the plugin.


## What does this PR do?

Adds a preflight check to the `register` phase of the plugin.
Implement a new method into the `PreflightCheck` class, reaching the remote ES to grab the version and checking against the local instance of `org.elasticsearch.Version`.
If the remote version is newer than local than raise a ConfigurationError.


## Why is it important/What is the impact to the user?

As a user that executes locally ingest pipelines I want to be sure to have a compability check before running, instead of crash at runtime if the local version of Ingest node codebase doesn't match with the version hosted on the remote Elasticsearch cluster.

## Checklist


## Author's Checklist

- [x] test locally with ES 8.10 should be ok (which match the ES ingest shipped the plugin as of 2023/11/24)
- [x] test locally with ES 8.11 must fail the version check
- [ ] added line in changelog and bumped version

## How to test this PR locally
- Download a version a of Elasticsearch with a minor greater than the minor contained in this plugin: https://github.com/elastic/logstash-filter-elastic_integration/blob/main/gradle.properties#L2. Eventually download a nightly from https://storage.googleapis.com/elastic-artifacts-staging/dra-info/index.html such as https://artifacts-snapshot.elastic.co/elasticsearch/8.12.0-c923b830/downloads/elasticsearch/elasticsearch-8.12.0-SNAPSHOT-darwin-aarch64.tar.gz
- build this plugin with 
```
export LOGSTASH_PATH=/tmp/logstash && export LOGSTASH_SOURCE=1
bundle install
bundle exec rake vendor
```
- In a Logstash update the `Gemfile` pointing to current checkout of the plugin and install it:
```
gem "logstash-filter-elastic_integration", :path => "/tmp/logstash_plugins/logstash-filter-elastic_integration"
/tmp/logstash/bin/logstash-plugin install --no-verify
```
- launch Elasticsearch and save the password that will appear on console after
```
Password for the elastic user 
```
- remove write permision to Elasticsearch's `http_ca.crt` file
```
chmod a-w /tmp/elastic_snapshot/config/certs/http_ca.crt
```
- create a pipeline file using this plugin with the password and http_ca.crt provided by Elasticsearch:
```
input {
  stdin {
    codec => json
  }
}

filter {
  elastic_integration {
    hosts => ["https://localhost:9200"]
    
    username => "elastic"
    password => "<password generated by ES>"

    geoip_database_directory => "/tmp/logstash/vendor/bundle/jruby/3.1.0/gems/logstash-filter-geoip-7.2.13-java/vendor/GeoLite2-City.mmdb"

    pipeline_name => "test_pipeline"

    ssl_enabled => true 
    ssl_certificate_authorities => ["/tmp/elasticsearch-8.12.0-SNAPSHOT/config/certs/http_ca.crt"]
  }
}

output {
  stdout {
    codec => rubydebug { metadata => true }
  }
}
```
 - verify that Logstash doesn't start because the versions doesn't match


## Related issues


- Closes #81 

## Use cases





## Logs

Logs that are printed when the versions doesn't. match:
```
[2023-11-24T14:08:41,735][ERROR][co.elastic.logstash.filters.elasticintegration.PreflightCheck][main] Exception checking version compatibility: illegal version format - snapshots are only supported until version 2.x
[2023-11-24T14:08:41,736][ERROR][logstash.javapipeline    ][main] Pipeline error {:pipeline_id=>"main", :exception=>#<LogStash::ConfigurationError: Preflight check failed: illegal version format - snapshots are only supported until version 2.x>, :backtrace=>["/Users/andrea/workspace/logstash_plugins/logstash-filter-elastic_integration/lib/logstash/filters/elastic_integration.rb:299:in `raise_config_error!'", "/Users/andrea/workspace/logstash_plugins/logstash-filter-elastic_integration/lib/logstash/filters/elastic_integration.rb:410:in `check_es_version!'", "/Users/andrea/workspace/logstash_plugins/logstash-filter-elastic_integration/lib/logstash/filters/elastic_integration.rb:375:in `perform_preflight_check!'", "/Users/andrea/workspace/logstash_plugins/logstash-filter-elastic_integration/lib/logstash/filters/elastic_integration.rb:134:in `register'", "org/logstash/config/ir/compiler/AbstractFilterDelegatorExt.java:75:in `register'", "/Users/andrea/workspace/logstash_andsel/logstash-core/lib/logstash/java_pipeline.rb:237:in `block in register_plugins'", "org/jruby/RubyArray.java:1989:in `each'", "/Users/andrea/workspace/logstash_andsel/logstash-core/lib/logstash/java_pipeline.rb:236:in `register_plugins'", "/Users/andrea/workspace/logstash_andsel/logstash-core/lib/logstash/java_pipeline.rb:611:in `maybe_setup_out_plugins'", "/Users/andrea/workspace/logstash_andsel/logstash-core/lib/logstash/java_pipeline.rb:249:in `start_workers'", "/Users/andrea/workspace/logstash_andsel/logstash-core/lib/logstash/java_pipeline.rb:194:in `run'", "/Users/andrea/workspace/logstash_andsel/logstash-core/lib/logstash/java_pipeline.rb:146:in `block in start'"], "pipeline.sources"=>["/Users/andrea/workspace/logstash_configs/filter_elastic_integration/pipeline.conf"], :thread=>"#<Thread:0x410da705 /Users/andrea/workspace/logstash_andsel/logstash-core/lib/logstash/java_pipeline.rb:134 run>"}
[2023-11-24T14:08:41,737][INFO ][logstash.javapipeline    ][main] Pipeline terminated {"pipeline.id"=>"main"}
[2023-11-24T14:08:41,747][ERROR][logstash.agent           ] Failed to execute action {:id=>:main, :action_type=>LogStash::ConvergeResult::FailedAction, :message=>"Could not execute action: PipelineAction::Create<main>, action_result: false", :backtrace=>nil}
[2023-11-24T14:08:41,751][INFO ][logstash.runner          ] Logstash shut down.
[2023-11-24T14:08:41,753][FATAL][org.logstash.Logstash    ] Logstash stopped processing because of an error: (SystemExit) exit
```
